bigframes/operations/type.py

```diff
@@ -22,54 +23,175 @@
+    def output_type(self, input_type: ExpressionType) -> ExpressionType:
+        return input_type
+
+
 @dataclasses.dataclass
-class InputType(OpTypeRule):
-    def output_type(self, *input_types: ExpressionType) -> ExpressionType:
-        assert len(input_types) == 1
-        return input_types[0]
+class TypePreserving(UnaryTypeSignature):
+    type_predicate: Callable[[ExpressionType], bool]
+    description: str
+
+    def output_type(self, input_type: ExpressionType) -> ExpressionType:
+        if not self.type_predicate(input_type):
+            raise TypeError(
+                f"Type {input_type} is not supported. Type must be {self.description}"
+            )
+        return input_type
+
+
+@dataclasses.dataclass
+class FixedOutputType(UnaryTypeSignature):
+    type_predicate: Callable[[ExpressionType], bool]
+    fixed_type: ExpressionType
+    description: str
+
+    def output_type(self, input_type: ExpressionType) -> ExpressionType:
+        if not self.type_predicate(input_type):
+            raise TypeError(
+                f"Type {input_type} is not supported. Type must be {self.description}"
+            )
+        return self.fixed_type
 
 
 @dataclasses.dataclass
-class RealNumeric(OpTypeRule):
-    def output_type(self, *input_types: ExpressionType) -> ExpressionType:
-        return functools.reduce(
-            lambda t1, t2: bigframes.dtypes.lcd_etype(t1, t2),
-            [*input_types, bigframes.dtypes.FLOAT_DTYPE],
-        )
+class UnaryRealNumeric(UnaryTypeSignature):
+    """Type signature for real-valued functions like exp, log, sin, tan."""
+
+    def output_type(self, type: ExpressionType) -> ExpressionType:
+        if not bigframes.dtypes.is_numeric(type):
+            raise TypeError(f"Type {type} is not numeric")
+        if type in (bigframes.dtypes.INT_DTYPE, bigframes.dtypes.BOOL_DTYPE):
+            return bigframes.dtypes.FLOAT_DTYPE
+        return type
 
 
 @dataclasses.dataclass
-class Supertype(OpTypeRule):
-    def output_type(self, *input_types: ExpressionType) -> ExpressionType:
-        return functools.reduce(
-            lambda t1, t2: bigframes.dtypes.lcd_etype(t1, t2), input_types
-        )
+class BinaryRealNumeric(BinaryTypeSignature):
+    """Type signature for real-valued functions like divide, arctan2, pow."""
+
+    def output_type(
+        self, left_type: ExpressionType, right_type: ExpressionType
+    ) -> ExpressionType:
+        if not bigframes.dtypes.is_numeric(left_type):
+            raise TypeError(f"Type {left_type} is not numeric")
+        if not bigframes.dtypes.is_numeric(right_type):
+            raise TypeError(f"Type {right_type} is not numeric")
+        lcd_type = bigframes.dtypes.lcd_etype(left_type, right_type)
+        if lcd_type == bigframes.dtypes.INT_DTYPE:
+            return bigframes.dtypes.FLOAT_DTYPE
+        return lcd_type
 
 
 @dataclasses.dataclass
-class Fixed(OpTypeRule):
-    out_type: ExpressionType
+class Numeric(BinaryTypeSignature):
```

Comment:
`BinaryNumeric`? Why we have both this and `BinaryRealNumeric`?

Reply:
The difference is whether integer inputs result in integer output for a float output. Some operations, such as logarithm, will produce float result even if input is an integer eg `log(3) ~= 0.47712`. Other operations, such as abs, preserve int type eg `abs(-2) == 2`